File: 0901-online-stock-span/0901-online-stock-span.py
class SpannerNode:
    
    def __init__(self, price: int = 0):
        self.price = price
        self.span = 1


class StockSpanner:

    def __init__(self):
        # self.spans: List[Tuple[int, int]] = []
        self.spans: List[SpannerNode] = []

    def next(self, price: int) -> int:
        quote = SpannerNode(price)
        
        while self.spans and price >= self.spans[-1].price:
            quote.span += self.spans.pop().span
        
        self.spans.append(quote)
        
        return quote.span


# Your StockSpanner object will be instantiated and called as such:
# obj = StockSpanner()
# param_1 = obj.next(price)

File: 0901-online-stock-span/0901-online-stock-span.go
type Quote struct {
    Price int
    Span  int
}
type StockSpanner struct {
    Quotes []Quote
}


func Constructor() StockSpanner {
    return StockSpanner{make([]Quote, 0)}
}


func (this *StockSpanner) Next(price int) int {
    span := 1
    
    for len(this.Quotes) > 0 && price >= this.Quotes[len(this.Quotes) - 1].Price {
        span += this.Quotes[len(this.Quotes) - 1].Span
        this.Quotes = this.Quotes[: len(this.Quotes) - 1]
    }
    
    this.Quotes = append(this.Quotes, Quote{Price: price, Span: span})
    
    return span
}


/**
 * Your StockSpanner object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Next(price);
 */
